csm_web/frontend/src/components/Course.js

```diff
@@ -0,0 +1,180 @@
+      <div id="course-section-selector">
+        <div id="course-section-controls">
+          <h2 className="course-title">{this.props.name}</h2>
+          <div id="day-selector">
+            {Object.keys(sections).map(day => (
+              <button
+                className={`day-btn ${day == currDay ? "active" : ""}`}
+                key={day}
+                onClick={() => this.setState({ day })}
+              >
+                {DAY_OF_WEEK_ABREVIATIONS[day]}
+              </button>
+            ))}
+          </div>
+          <label id="show-unavailable-toggle">
+            <input
+              type="checkbox"
+              checked={this.state.showUnavailable}
+              onChange={({ target: { checked } }) => this.setState({ showUnavailable: checked })}
+            />
+            Show unavailable
+          </label>
+        </div>
+        <div id="course-section-list">
+          {sections[currDay]
+            .filter(({ numStudentsEnrolled, capacity }) => this.state.showUnavailable || numStudentsEnrolled < capacity)
+            .map(section => (
+              <SectionCard key={section.id} {...section} />
+            ))}
+        </div>
+      </div>
+    );
+  }
+}
+
+class SectionCard extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { showModal: false, enrollmentSuccessful: undefined, errorMessage: "" };
+    this.enroll = this.enroll.bind(this);
+    this.closeModal = this.closeModal.bind(this);
+    this.modalContents = this.modalContents.bind(this);
+  }
+
+  static propTypes = {
+    id: PropTypes.number.isRequired,
+    location: PropTypes.string,
+    time: PropTypes.string.isRequired,
+    mentor: PropTypes.shape({ name: PropTypes.string.isRequired }),
+    numStudentsEnrolled: PropTypes.number.isRequired,
+    capacity: PropTypes.number.isRequired
+  };
+
+  enroll() {
+    fetchWithMethod(`sections/${this.props.id}/students/`, HTTP_METHODS.PUT)
+      .then(response => {
+        if (response.ok) {
+          this.setState({
+            showModal: true,
+            enrollmentSuccessful: true
+          });
+        } else {
+          response.json().then(({ detail }) =>
+            this.setState({
+              showModal: true,
+              enrollmentSuccessful: false,
+              errorMessage: detail
+            })
+          );
+        }
+      })
+      .catch(error =>
+        this.setState({
+          showModal: true,
+          enrollmentSuccessful: false,
+          errorMessage: String(error)
+        })
+      );
+  }
+
+  closeModal() {
+    this.setState({ showModal: false, modalContents: null });
+  }
+
+  modalContents() {
+    const iconWidth = "10em";
+    const iconHeight = "10em";
+    if (this.state.enrollmentSuccessful) {
+      return (
+        <React.Fragment>
+          <CheckCircle color="green" height={iconHeight} width={iconWidth} />
+          <h3>Successfully enrolled</h3>
+          <button className="modal-btn" modalClose>
```

Comment:
From the console: "React does not recognize the `modalClose` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `modalclose` instead. If you accidentally passed it from a parent component, remove it from the DOM element."

After messing around a bit, it seems like `modalclose` or `modalclose="true"` both still incur a warning despite the console's advisories; `modalclose={1}` worked though.

Reply:
Hmm. So the thing is we _don't_ want `modalClose` to actually appear on the DOM element, things are working as intended. `modalClose` is just used by the `Modal` component to indicate that this element should close the modal when it is clicked (in addition to performing its own `onClick` event if there is one). We could do something hacky like add a special class name to the element instead of using the custom `modalClose` prop, but then we definitely leak this into the DOM, all for the sake of suppressing a warning. Not sure what the best compromise is here.

Reply:
Is it a problem that it gets leaked onto the DOM? I was under the impression that this is the kind of thing that class names should be used for.

Reply:
It's not a problem per-say, it just strikes me as a little odd to have a CSS class on an element that doesn't actually apply any styling, but is used purely for JS logic.

Here's an alternative implementation I just thought of. Inside Modal.js we define
```javascript
export function ModalCloser({children}) {
    return <React.Fragment>{children}</React.Fragment>;
}
```

And you just wrap anything that should close the modal in this element instead of using the `modalClose` prop. Then inside the `Modal` component you have `child.type == ModalCloser` instead of `Object.prototype.hasOwnProperty.call(child.props, "modalClose")`

Reply:
sure, that makes sense

Reply:
Wait actually that has issues, it means you can't nest anything that will close the Modal inside anything else.

Reply:
Resolved using that idea with React context